src/vcpkg/base/files.cpp

```diff
@@ -1682,6 +1824,206 @@ namespace vcpkg
         {
             return get_regular_files_impl<stdfs::directory_iterator>(dir, ec);
         }
+#else // ^^^ _WIN32 // !_WIN32 vvv
+#if defined(_DIRENT_HAVE_D_TYPE)
+#define VCPKG_GET_D_TYPE(dirent) ((dirent)->d_type)
+#define VCPKG_DT_UNKNOWN DT_UNKNOWN
+#define VCPKG_DT_REG DT_REG
+#define VCPKG_DT_DIR DT_DIR
```

Comment:
I really don't like that these are macros...

```suggestion
    enum class PosixDType {
        Unknown = DT_UNKNOWN,
        Regular = DT_REG,
        Directory = DT_DIR,
    };
    PosixDType get_d_type(const dirent* d) noexcept {
        return d->d_type;
    }
```